test/samplers/test_samplers.py

```diff
@@ -1,21 +1,66 @@
 import contextlib
 import random
 import re
+from copy import copy
+from functools import partial
 
 import pytest
 import torch
 
 from torchcodec import FrameBatch
 from torchcodec.decoders import VideoDecoder
-from torchcodec.samplers import clips_at_random_indices, clips_at_regular_indices
-from torchcodec.samplers._implem import _build_all_clips_indices, _POLICY_FUNCTIONS
+from torchcodec.samplers import (
+    clips_at_random_indices,
+    clips_at_regular_indices,
+    clips_at_regular_timestamps,
+)
+from torchcodec.samplers._implem import (
+    _build_all_clips_indices,
+    _build_all_clips_timestamps,
+    _POLICY_FUNCTIONS,
+)
 
 from ..utils import assert_tensor_equal, NASA_VIDEO
 
 
+def _assert_output_type_and_shapes(
+    video, clips, expected_num_clips, num_frames_per_clip
+):
+    assert isinstance(clips, list)
+    assert len(clips) == expected_num_clips
+    assert all(isinstance(clip, FrameBatch) for clip in clips)
+    expected_clip_data_shape = (
+        num_frames_per_clip,
+        3,
+        video.height,
+        video.width,
+    )
+    assert all(clip.data.shape == expected_clip_data_shape for clip in clips)
+
+
+def _assert_regular_sampler(clips, expected_seconds_between_clip_starts=None):
+    # assert regular spacing between sampled clips
+    seconds_between_clip_starts = torch.tensor(
+        [clip.pts_seconds[0] for clip in clips]
+    ).diff()
+    for diff in seconds_between_clip_starts:
+        # Note: need approximate check as actual diff values typically look like
+        # [3.2032, 3.2366, 3.2366, 3.2366]
+        assert diff == pytest.approx(seconds_between_clip_starts[0], abs=0.05)
+
+    if expected_seconds_between_clip_starts is not None:
+        # This can only be asserted with the time-based sampler, where
+        # seconds_between_clip_starts is known since it's passed by the user.
+        torch.testing.assert_close(
+            diff, expected_seconds_between_clip_starts, atol=0.01, rtol=0
+        )
+
+    assert (diff > 0).all()  # Also assert clips are sorted by start time
+
```

Comment:
Is line 51 supposed to be indented one level? As written, I think it will only apply to the last value `diff` takes on in the loop above. (I didn't even realized iteration variables remained in scope after their loop in Python until I tested it just now.)

I'm also confused by line 58 - like line 51 above, I would expect it to be indented one level and just be `assert diff > 0`. As written, though, I'm very confused how it passes, as I thought `(diff > 0)` would become a boolean, and booleans don't have an `all()` method.

Is there something obvious I'm missing?

Reply:
Good catch, thank you!

I botched it while I did the refactoring *and* the original code wasn't even correct either. 

> As written, though, I'm very confused how it passes, as I thought (diff > 0) would become a boolean, and booleans don't have an all() method

`diff` was a tensor so the code didn't fail, but my assertion was still asserting the wrong thing.